src/lib/pubkey/pcurves/pcurves_impl.h

```diff
@@ -0,0 +1,816 @@
+         } else {
+            static_assert(8 * Self::BYTES == Self::BITS);
+
+            const std::array<W, N> hi = bytes_to_words<W, N, BYTES>(&padded_bytes[0]);
+            const std::array<W, N> lo = bytes_to_words<W, N, BYTES>(&padded_bytes[BYTES]);
+            return Self(hi) * R3 + Self(lo) * R2;
+         }
+      }
+
+      static constexpr Self random(RandomNumberGenerator& rng) {
+         std::array<uint8_t, Self::BYTES> buf;
+         for(;;) {
+            rng.randomize(buf.data(), buf.size());
+            if(auto v = Self::deserialize(buf)) {
+               return v;
+            }
+         }
+      }
+
+      template <size_t N>
+      static consteval Self constant(StringLiteral<N> S) {
+         return Self::constant(S.value);
+      }
+
+      template <size_t N>
+      static consteval Self constant(const char (&s)[N]) {
+         const auto v = hex_to_words<W>(s);
+         return Self(v) * R2;
+      }
+
+      static consteval Self constant(int8_t x) {
+         std::array<W, 1> v;
+         v[0] = (x >= 0) ? x : -x;
+         auto s = Self(v) * R2;
+         return (x >= 0) ? s : s.negate();
+      }
+
+   private:
+      constexpr const std::array<W, N>& value() const { return m_val; }
+
+      constexpr const W* data() const { return m_val.data(); }
+
+      template <size_t S>
+      constexpr MontgomeryInteger(std::array<W, S> w) : m_val({}) {
+         static_assert(S <= N);
+         for(size_t i = 0; i != S; ++i) {
+            m_val[i] = w[i];
+         }
+      }
+
+      std::array<W, N> m_val;
+};
+
+template <typename FieldElement>
+class AffineCurvePoint {
+   public:
+      static const constinit size_t BYTES = 1 + 2 * FieldElement::BYTES;
+      static const constinit size_t COMPRESSED_BYTES = 1 + FieldElement::BYTES;
+
+      typedef AffineCurvePoint<FieldElement> Self;
+
+      constexpr AffineCurvePoint(const FieldElement& x, const FieldElement& y) : m_x(x), m_y(y) {}
+
+      constexpr AffineCurvePoint() : m_x(FieldElement::zero()), m_y(FieldElement::zero()) {}
+
+      static constexpr Self identity() { return Self(FieldElement::zero(), FieldElement::zero()); }
+
+      constexpr bool is_identity() const { return m_x.is_zero() && m_y.is_zero(); }
+
+      AffineCurvePoint(const Self& other) = default;
+      AffineCurvePoint(Self&& other) = default;
+      AffineCurvePoint& operator=(const Self& other) = default;
+      AffineCurvePoint& operator=(Self&& other) = default;
+
+      constexpr Self negate() const { return Self(m_x, m_y.negate()); }
+
+      std::vector<uint8_t> serialize_to_vec() const {
+         const auto b = this->serialize();
+         return std::vector(b.begin(), b.end());
+      }
+
+      constexpr std::array<uint8_t, Self::BYTES> serialize() const {
+         std::array<uint8_t, Self::BYTES> r = {};
+         BufferStuffer pack(r);
+         pack.append(0x04);
+         pack.append(m_x.serialize());
+         pack.append(m_y.serialize());
+         return r;
+      }
```

Comment:
It would be nice if we could use the `concat` helpers for this. Currently, that doesn't work, because `concat` requires the output type to have `.insert()`. But, technically, we could extend it to be used like that:

```suggestion
      constexpr std::array<uint8_t, Self::BYTES> serialize() const {
         return concat_as<std::array<uint8_t, Self::BYTES>(
            store_le<uint8_t>(0x04), // store_le is just to make it look like a byte-array
            m_x.serialize(),
            m_y.serialize());
      }
```

This could then also (statically) assert that the output array was filled entirely.

Reply:
#3994 is work towards this. It makes `concat()` `constexpr` and also automagically inferring the output array size. So this should actually work as:

```suggestion
      constexpr std::array<uint8_t, Self::BYTES> serialize() const {
         return concat(
            store_le<uint8_t>(0x04), // store_le is just to make it look like a byte-array
            m_x.serialize(),
            m_y.serialize());
      }
```

Reply:
Something seems to be missing still

```
build/include/internal/botan/internal/loadstor.h: In instantiation of ‘constexpr auto Botan::store_le(ParamTs&& ...) [with ModifierT = unsigned char; ParamTs = {int}]’:
build/include/internal/botan/internal/pcurves_impl.h:351:30:   required from here
build/include/internal/botan/internal/loadstor.h:699:67: error: no matching function for call to ‘store_any<Botan::detail::Endianness::Little, unsigned char>(int)’
  699 |    return detail::store_any<detail::Endianness::Little, ModifierT>(std::forward<ParamTs>(params)...);
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
With move to span arg we don’t need concat anymore